forgit 23.04.0 (new formula)

carlfriedrich opened the pull request:
Fixes https://github.com/wfxr/forgit/issues/194

<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

chenrui333 commented on Formula/forgit.rb:
```diff
@@ -0,0 +1,17 @@
+  end
+
+  test do
+    system "which", "git-forgit"
```
Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests.

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

carlfriedrich replied:
Hi @chenrui333, thanks for your feedback. `forgit` is a fully interactive tool, so I actually don't know how to test any functionality of it without user input. Do you have an idea?

carlfriedrich replied:
To explain it even further: `forgit` is an interactive frontend for `git`. So the actual *functionality* that coud be tested is just functionality of `git`. Everything `forgit` does is just user presentation. How could we test this?

carlfriedrich replied:
Nevermind, I think I found a useful testcase. `git forgit add <file>` bypasses the interactive selector while still checking if forgit is correctly installed, because otherwise `git` would complain:

`git: 'forgit' is not a git command. See 'git --help'.`

So this should be a valid test IMO. Let me know if that works for you.

carlocab commented on Formula/forgit.rb:
```diff
@@ -0,0 +1,17 @@
+  depends_on "fzf"
+
+  def install
+    bin.install "bin/git-forgit" => "git-forgit"
```
This should probably be enough:
```suggestion
    bin.install "bin/git-forgit"
```

carlocab replied:
Or, better:
```suggestion
    prefix.install "bin"
    bash_completion.install "completions/git-forgit.bash"
    pkgshare.install "forgit.plugin.zsh"
```

carlfriedrich replied:
@carlocab I also had to adjust `pkgshare.install "forgit.plugin.zsh"` to `prefix.install "forgit.plugin.zsh"` because the shell plugin assumes that the forgit script is located in a subdir `bin` relative to the plugin location.

carlfriedrich commented:
@carlocab @chenrui333 I updated the forumla according to your reviews. I also added a message to the user about how to use the forgit installation. I assume that adding a line to the user's `.zshrc` automatically is not a task homebrew should do?

carlocab commented on Formula/forgit.rb:
```diff
@@ -0,0 +1,28 @@
+  def install
+    bin.install "bin/git-forgit"
+    bash_completion.install "completions/git-forgit.bash"
+    prefix.install "forgit.plugin.zsh"
```
Let's use `pkgshare` since that's where other formulae install Zsh plugins.
```suggestion
    inreplace "forgit.plugin.zsh", 'FORGIT="$INSTALL_DIR', "FORGIT=\"#{opt_prefix}"
    pkgshare.install "forgit.plugin.zsh"
    pkgshare.install_symlink "forgit.plugin.zsh" => "forgit.plugin.sh"
```

carlfriedrich replied:
@carlocab Great, thanks a lot for the feedback. I updated the branch with your sugesstions.

carlocab commented on Formula/forgit.rb:
```diff
@@ -0,0 +1,28 @@
+    prefix.install "forgit.plugin.zsh"
+  end
+
+  def post_install
+    ohai "NOTE: forgit can be used as subcommand of git (e.g. 'git forgit log')."
+    ohai "If you like to have zsh/bash shell aliases (e.g. 'glo'), source the shell plugin:"
+    ohai "  source #{prefix}/forgit.plugin.zsh"
+    ohai "Add the line to your .zshrc/.bashrc to make this permanent."
+  end
```
```suggestion
  def caveats
    <<~EOS
      A shell plugin has been installed at
        #{opt_pkgshare}/forgit.plugin.zsh
        #{opt_pkgshare}/forgit.plugin.sh
    EOS
  end
```

carlfriedrich replied:
@carlocab Good idea, thanks a lot. I updated the branch accordingly.

carlocab reviewed (approved):
Thanks!
